### PR TITLE
Add array reduce tests to address issue #8

### PR DESCRIPTION
**Pull request by jhdavis8**
These are a pair of tests on array reductions that come from issue #8 . They are attempting to perform array operations in a set of nested loops on the device, reducing to array elements.

The two versions (v1 and v2) differ in how they handle the reduction. The first test uses the default OpenMP addition reduction, while the second declares a custom reduction operation, which should get the same results.

I converted these codes into tests by just performing the same math and the host and on the device and then checking the results, on some dummy generated data, since none was provided.

Right now, the only test that is passing is the first verion on gfortran. The second test will not compile on gfortran, complaining about a type conversion in the arrayadd module. I suspect that the target reduction might not be written correctly as I had to correct other errors as well, so it would be good to get some more opinions on that. On IBM compilers both tests fail, as mentioned in the issue.

### CODE REVIEW

File: tests/4.5/application_kernels/array_reduce_v1.F90 [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  OMPVV_TEST_OFFLOADING
+  OMPVV_TEST_SHARED_ENVIRONMENT
+
+  OMPVV_TEST_VERBOSE(test_array_reduce() .ne. 0)
+
+  OMPVV_REPORT_AND_RETURN()
+CONTAINS
+  INTEGER FUNCTION test_array_reduce()
+
+    ! xl : mpif90 -O3 -qstrict -qsmp=omp -qoffload array_reduce_v1.F90 -o array_reduce.x
```
Comment (josemonsalve2) [inline tests/4.5/application_kernels/array_reduce_v1.F90]: what is this comment for?

File: tests/4.5/application_kernels/array_reduce_v1.F90 [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+          DO yp = 1, my
+             vx(x, z, yp) = yp
+          END DO
+       END DO
+    END DO
+
+    !$omp target data map(to:vx) map(from:rrs)
+    !$omp target
+    rrs(:,:) = 0.
+    !$omp end target
```
Comment (josemonsalve2) [inline tests/4.5/application_kernels/array_reduce_v1.F90]: Why initialize it in the device?

File: tests/4.5/application_kernels/array_reduce_v2.F90 [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  OMPVV_TEST_OFFLOADING
+  OMPVV_TEST_SHARED_ENVIRONMENT
+
+  OMPVV_TEST_VERBOSE(test_array_reduce() .ne. 0)
+
+  OMPVV_REPORT_AND_RETURN()
+CONTAINS
+  INTEGER FUNCTION test_array_reduce()
+
+    ! xl : mpif90 -O3 -qstrict -qsmp=omp -qoffload array_reduce.F90 -o array_reduce.x
```
Comment (josemonsalve2) [inline tests/4.5/application_kernels/array_reduce_v2.F90]: what is this for?

File: tests/4.5/application_kernels/array_reduce_v2.F90 [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    !$omp target data map(to:vx) map(from:rrs)
+    !$omp target
+    rrs(:,:)%x=0.
+    !$omp end target
+
+    ! separation in x
+    !$omp target teams distribute parallel do collapse(4) default(none) &
+    !$omp private(yp, z, x, x1, ir, term1, term2, sqterm) shared(my, &
+    !$omp nz, nx, nxh, inc, norder, vx) &
+    !$omp reduction(+:rrs)
```
Comment (josemonsalve2) [inline tests/4.5/application_kernels/array_reduce_v2.F90]: Really interesting test! 

I wonder if this is supported by 4.5

File: tests/4.5/application_kernels/array_reduce_v1.F90 (hunk 1 above)
Comment (jhdavis8) [reply tests/4.5/application_kernels/array_reduce_v1.F90]: I believe this comment was added to record how they compiled the test. I will remove it.

File: tests/4.5/application_kernels/array_reduce_v1.F90 (hunk 2 above)
Comment (jhdavis8) [reply tests/4.5/application_kernels/array_reduce_v1.F90]: I'm not sure actually, I'll change it

File: tests/4.5/application_kernels/array_reduce_v2.F90 [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+! compares the result. The reduction operator is a custom declared add
+! operator over reals. The reduction occurs within a four-level collapsed
+! loop.
+!
+!//===----------------------------------------------------------------------===//
+
+#include "ompvv.F90"
+#define N 768
+
+MODULE arrayadd_module
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/array_reduce_v2.F90]: Try to take out the reduction and see if the error stays

File: tests/4.5/application_kernels/array_reduce_v1.F90 [hunk 6]
```diff
@@ -0,0 +1,114 @@
+!===--- array_reduce_v1.F90 ------------------------------------------------===//
+!
+! OpenMP API Version 4.5 Nov 2015
+!
+! This test conducts an array reduction on both the host and device and
+! compares the result. The reduction operator is the default OpenMP add
+! operator over reals. The reduction occurs within a four-level collapsed
+! loop.
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/array_reduce_v1.F90]: Add comment describing where it came from

**Comment by jhdavis8** [comment]
Update on the status of this test: only v1 is passing, and only on gfortran.

array_reduce_v1 xlf error: 1587-175 The underlying GPU runtime reported the following error "an illegal memory access was encountered". 

array_reduce_v2 gfortran error: /autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/4.5/application_kernels/array_reduce_v2.F90:50:6:

   USE arrayadd_module
      1
Error: Can't convert REAL(4) to TYPE(myreal) at (1)

array_reduce_v2 xlf error: "/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/4.5/application_kernels/array_reduce_v2.F90", 1516-044 (S) A conversion from type REAL*4 is not permitted.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/4.5/application_kernels/array_reduce_v2.F90", 1516-064 (S) Operands to the + operator must be compatible with the intrinsic uses of the operator, or 
with a specific interface within an accessible defined operator generic interface.

**Comment by jhdavis8** [comment]
Update: neither tests pass any compiler on Summit. It might be a good idea to contact the authors to see if the errors make sense, since I am unsure if I fully understand this code.

File: tests/4.5/application_kernels/array_reduce_v1.F90 [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+#define N 768
+
+PROGRAM array_reduce_v1
+  USE iso_fortran_env
+  USE ompvv_lib
+  USE omp_lib
+  implicit none
+
+  OMPVV_TEST_OFFLOADING
+  OMPVV_TEST_SHARED_ENVIRONMENT
```
Comment (spophale) [inline tests/4.5/application_kernels/array_reduce_v1.F90:22]: remove.

**Comment by jhdavis8** [comment]
I am closing this PR for now since it has been sitting on the docket for quite a while with no progress made. This test could be resurrected with some time, but it should try to use a simpler example/testing strategy.

**Pull request closed without merging**